contributors/design-proposals/network/pod-resolv-conf.md

```diff
@@ -0,0 +1,224 @@
+      options:
+      - "ndots:2"
+```
+
+Given the following host `/etc/resolv.conf`:
+
+```bash
+nameserver 1.2.3.4
+search foo.com bar.com
+options ndots:1
+```
+
+The pod will get the following `/etc/resolv.conf`:
+
+```bash
+nameserver 10.240.0.10
+# Comments only for explication purposes.
+#
+#      $(NAMESPACE).svc.$CLUSTER
+#      |                     my.dns.search.suffix
+#      |                     |                    $(HOST)
+#      |                     |                    |
+#      V                     V                    V
+search ns1.svc.cluster.local my.dns.search.suffix foo.com bar.com
+# Populated from custom.options
+options ndots:2
+```
+
+## More examples
+
+The following is a Pod dnsParams that only contains the host search paths:
+
+```yaml
+dnsParams:
+  custom:
+    searchPaths:
+    - $HOST
+```
+
+Override `ndots` and add custom search path. Note that overriding the ndot may
+break the functionality of some of the search paths the
+
+```yaml
+dnsParams:
+  custom:
+    searchPaths:
+    - my.custom.suffix
+    - $HOST
+    options:
+    - "ndots:3"
+```
+
+# API changes
+
+```go
+type PodSpec struct {
+	...
+	DNSPolicy string
+	DNSParams *PodDNSParams
+    ...
+}
+
+type PodDNSParams struct {
+ 	Custom PodDNSParamsCustom
+}
+
+type PodDNSParamsCustom struct {
+	Search []string
+ 	Options []string
+}
+
+// This will not appear in types.go but is here for explication purposes.
+type DNSParamsSubstitution string
+const (
+	DNSParamsSearchPathNamespace = "$(NAMESPACE)"
+	DNSParamsSearchPathClusterDomain = "$(CLUSTER)"
+	DNSParamsSearchPathHostPaths = "$(HOST)"
+)
+```
+
+## Semantics
+### searchPath
+
+If `dnsPolicy: Custom` is used, then the `search` line will be constructed from
+the entries listed in `dnsParams.custom.search`:
+
+```go
+// Note: pseudocode does not include input validation.
+func SearchPath(params *PodDNSParams) []string {
+	var searchPaths []string
+	for _, entry := range params.SearchPaths {
+		if entry == "$HOST" {
+			searchPaths = append(searchPaths, HostSearchPaths...)
+			break
+		}
+		searchPaths = append(searchPaths, expand(entry))
+	}
+	return searchPaths
+}
+
+func expand(s string) string {
+	labels := strings.Split(s, ".")
+	for i := range labels {
+		if labels[0] == "$" {
+			labels[i] = Substitute(labels[i])
+		}
+	}
+	return strings.Join(labels, ".")
+}
+
+var LabelSubstitutions map[string]string
+
+func Substitute(l label) string {
+  if s, ok := LabelSubstitutions[l]; ok {
+    return s
+  }
+  return l
+}
+```
+
+#### Substitutions
+
+`Substitute` will replace DNS labels that begin with `$` with values for the
+given Pod. Note: the DNS label MUST be an EXACT string match. E.g. if the pod
+namespace is `my-ns`, `abc$(NAMESPACE)1234` will NOT be expanded to
+`abcmy-ns1234`, but will be an invalid configuration.
+
+| Substitution | Description |
+| ----   | ---- |
+| `$(NAMESPACE)` | Namespace of the Pod |
+| `$(CLUSTER)` | Kubernetes cluster domain (e.g. `cluster.local`) |
```

Comment:
The moment we add this people will ask for it in other places.  This should be part of the downward API if we want to add it.

Reply:
And it needs a better name, it's not `CLUSTER` it's `CLUSTER_SUBDOMAIN` or similar.

Reply:
Making it part of downward API is an interesting idea.  Do we have other places where downward API pulls values from thin air?  ISTR most of them are pulled from pod fields.  Namespace could be metadata.namespace (if it isn't already), but we have no field for cluster suffix.

This proposal was doing a limited-set of substitutions outside of downward API because, as we know, the downward API requires more structure than simple templating.  These were not intended to be generic env-vars that expand, just a few magic names in this particular piece of the API. 
 It's hacky, and I figured it would draw attention, but I hoped we could keep it simple (since there's no risk of collision - we control all the variables, unlike `env`).

We could make this a generic env expansion, which would require users to touch both `env` and `dnsPolicy`.  That feels like overkill.

Reply:
The last time I looked at making this part of the downward API, this seems the first in terms of values being passed that are not part of a declared spec.